File: sem iii/NM/Nm lab/Secant_lab.c
#include <stdio.h>
#include <math.h>

// f(x) = 3x - cos(x) - 1
double f(double x) {
    return 3*x - cos(x) - 1;
}

int main() {
    double xprev, x, x_new, error, eps;
    int itr, i;

    printf("Program to find root using Secant Method\n");
    printf("Enter the initial guess: ");
    scanf("%lf", &xprev);
    printf("Enter the next initial guess: ");
    scanf("%lf", &x);
    printf("Enter the error tolerance: ");
    scanf("%lf", &eps);
    printf("Enter the number of iterations: ");
    scanf("%d", &itr);

    for (i = 0; i < itr; i++) {
        x_new = x - (f(x) * (x - xprev)) / (f(x) - f(xprev));
        error = fabs((x_new - x) / x_new);

        printf("Iteration %d: Root = %.4lf  Error = %.4lf\n",
               i+1, x_new, error);

        if (error < eps)
            break;

        xprev = x;
        x = x_new;
    }

    printf("\nThe root of the function is %.4lf\n", x_new);
    return 0;
}

File: sem iii/NM/Nm lab/ho/Secant_lab.py
import math

# Lab 3 - Secant Method
# f(x) = 3x - cos(x) - 1

# define function f(x)
def f(x):
    return 3*x - math.cos(x) - 1


# Secant Method function
def Secant(xi, xprev, itr, eps):
    for i in range(itr):
        fxi = f(xi)
        fxprev = f(xprev)

        # Secant formula
        x_new = xi - (fxi * (xi - xprev)) / (fxi - fxprev)

        # relative approximate error
        error = abs((x_new - xi) / x_new)

        print(f"The root estimate is: {x_new:.4f} and the error is: {error:.4f}")

        # stopping condition
        if error < eps:
            return x_new
        else:
            xprev = xi
            xi = x_new

    return xi


if __name__ == '__main__':
    print("Program to find the root of the given function using Secant method.")
    xprev = float(input("Enter the initial guess: "))
    x = float(input("Enter the next initial guess: "))
    eps = float(input("Enter the error tolerance value: "))
    itr = int(input("Enter the number of iterations: "))

    root = Secant(x, xprev, itr, eps)
    print(f"\nThe root of the function is {root:.4f}")
